contracts/ERC1155NftCustom.sol

```diff
@@ -0,0 +1,155 @@
+    }
+
+    function setURI(string memory _newURI) public onlyRole(MINTER_ROLE) {
+        _setURI(_newURI);
+    }
+
+    function updateTokenUri(uint256 _tokenId, string memory _newUri, bool _isFreezeTokenUri)
+    public
+    onlyRole(MINTER_ROLE) {
+        require(_exists(_tokenId), "NFT: update URI query for nonexistent token");
+        require(isFreezeTokenUris == false, "NFT: Token uris are frozen globally");
+        require(freezeTokenUris[_tokenId] != true, "NFT: Token is frozen");
+        require(_isFreezeTokenUri || (bytes(_newUri).length != 0), "NFT: Either _newUri or _isFreezeTokenUri=true required");
+
+        if (bytes(_newUri).length != 0) {
+            require(keccak256(bytes(_tokenURIs[_tokenId])) != keccak256(bytes(string(abi.encodePacked(_newUri)))), "NFT: New token URI is same as updated");
+            _tokenURIs[_tokenId] = _newUri;
+            emit URI(_newUri, _tokenId);
+        }
+        if (_isFreezeTokenUri) {
+            freezeTokenUris[_tokenId] = true;
+            emit PermanentURI(_tokenURIs[_tokenId], _tokenId);
+        }
+    }
+
+
+    function update(string memory _newBaseURI, bool _freezeAllTokenUris)
+    public
+    onlyRole(MINTER_ROLE) {
+        require(isFreezeTokenUris == false, "NFT: Token uris are already frozen");
+        baseURI = _newBaseURI;
+        if (_freezeAllTokenUris) {
+            freezeAllTokenUris();
+        }
+    }
+
+    function freezeAllTokenUris()
+    public
+    onlyRole(MINTER_ROLE) {
+        require(isFreezeTokenUris == false, "NFT: Token uris are already frozen");
+        isFreezeTokenUris = true;
+
+        emit PermanentURIGlobal();
+    }
+
+    function totalSupply (uint256 _id) public view returns (uint256) {
+        return tokenSupply[_id];
+    }
+
+    function uri(uint256 _id) public override view returns (string memory) {
+        if (bytes(_tokenURIs[_id]).length > 0) {
+            if (bytes(baseURI).length > 0) {
+                return string(abi.encodePacked(baseURI, _tokenURIs[_id]));
+            } else {
+                return _tokenURIs[_id];
+            }
+        } else {
+            return super.uri(_id);
+        }
+    }
+
+    function mintToCaller( address account, uint256 id, uint256 amount, string memory uri) public onlyRole(MINTER_ROLE) {
+        require(!_exists(id), "NFT: token already minted");
+        if (bytes(uri).length > 0) {
+            _tokenURIs[id] = uri;
+            emit URI(uri, id);
+        }
+        _mint(account, id, amount, "");
+        tokenSupply[id] += amount;
+    }
+
+    function mintToCallerBatch(
+        address to,
+        uint256[] memory ids,
+        uint256[] memory amounts,
+        string[] memory uris
+    ) public onlyRole(MINTER_ROLE) {
+        for (uint256 i = 0; i < ids.length; i++) {
+            require(!_exists(ids[i]), "NFT: one of tokens are already minted");
+            tokenSupply[ids[i]] += amounts[i];
+            if (bytes(uris[i]).length > 0) {
```

Comment:
if uris array has less elements than ids does this crash btw?

Reply:
Thanks for pointing this out, will check

Reply:
Checked, added this case to test  too